internal/infrastructure/elastic_search/query_device_features.go

```diff
@@ -0,0 +1,141 @@
+package elasticsearch
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+type DeviceFeatures struct {
+	BatteryVoltage struct {
+		DocumentCount int64 `json:"doc_count"`
+	}
+	FuelPercentRemaining struct {
+		DocumentCount int64 `json:"doc_count"`
+	}
+	Odometer struct {
+		DocumentCount int64 `json:"doc_count"`
+	}
+	Oil struct {
+		DocumentCount int64 `json:"doc_count"`
+	}
+	Soc struct {
+		DocumentCount int64 `json:"doc_count"`
+	}
+	Speed struct {
+		DocumentCount int64 `json:"doc_count"`
+	}
+	Tires struct {
+		DocumentCount int64 `json:"doc_count"`
+	}
+}
+
+type DeviceFeaturesResp struct {
+	Aggregations struct {
+		Features struct {
+			SumOtherDocCount int64 `json:"sum_other_doc_count"`
+			Buckets          []struct {
+				Key               string `json:"key"`
+				DocumentCount     int64  `json:"doc_count"`
+				DeviceDefinitions struct {
+					SumOtherDocCount int64 `json:"sum_other_doc_count"`
+					Buckets          []struct {
+						Key           string `json:"key"`
+						DocumentCount int64  `json:"doc_count"`
+						Features      struct {
+							Buckets map[string]map[string]int
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+func (d *ElasticSearch) GetDeviceFeatures(envName string) (DeviceFeaturesResp, error) {
+	url := fmt.Sprintf("%s/device-status-%s*/_search", d.BaseURL, envName)
+	body := `
+	{
+		"size": 0,
+		"query": {                                                                    
+		  "bool": {
+			"must": [
+			  {
+				"exists": {
+				  "field": "data.deviceDefinitionId"
+				}
+			  }
+			]
+		  }
+		},
+		"aggs": {
+		  "features": {
+			"terms": {
+			  "field": "source",
+			  "size": 1000
+			},
+			"aggs": {
+			  "deviceDefinitions": {
+				"terms": {
+				  "field": "data.deviceDefinitionId",
+				  "size": 10000
+				},
+				"aggs": {
+			  "features":{
+				"filters": {
+				  "filters": {
+					"odometer":{
+					  "exists": {
+						"field": "data.odometer"
+					  }
+					},
```

Comment:
Here's the interesting part, having combined what might have been two PRs: I think you want to construct this section using the `device_features` table. So now it's (using pseudocode)
```
for row in device_integrations:
    filters[row.feature_key] = {"exists": {"field": "data." + row.elastic_field}}
```

Reply:
Yes that makes sense..